[Varint] Implement ShreddingState::AllNull variant

codephage2020 opened the pull request:
# Which issue does this PR close?

- Closes #8088 .

# What changes are included in this PR?

handles the "all null" case

# Are these changes tested?

Yes.

# Are there any user-facing changes?

no.

alamb reviewed (commented):
Thank you @codephage2020  -- this looks great -- I had one corner case question

codephage2020 commented on parquet-variant-compute/src/variant_array.rs:
```diff
@@ -488,4 +495,50 @@ mod test {
+    #[test]
+    fn all_null_variant_array_construction() {
+        let metadata = BinaryViewArray::from(vec![b"test" as &[u8]; 3]);
+        let nulls = NullBuffer::from(vec![false, false, false]); // all null
```
Oh, I got it. I will add a test for the case.

codephage2020 replied:
I added a comprehensive test that demonstrates when a value field exists in the schema but contains all null  values, it correctly remains in the Unshredded state rather than AllNull.

scovich commented on parquet-variant-compute/src/variant_get/output/primitive.rs:
```diff
@@ -155,6 +156,19 @@ impl<'a, T: ArrowPrimitiveVariant> OutputBuilder for PrimitiveOutputBuilder<'a,
+        _metadata: &BinaryViewArray,
+    ) -> Result<ArrayRef> {
+        // For all-null case, create a primitive array with all null values
+        let nulls = NullBuffer::from(vec![false; variant_array.len()]);
+        let values = vec![T::default_value(); variant_array.len()];
+        let array = PrimitiveArray::<T>::new(values.into(), Some(nulls))
+            .with_data_type(self.as_type.data_type().clone());
+        Ok(Arc::new(array))
```
```suggestion
        Ok(Arc::new(new_null_array(self.as_type.data_type(), variant_array.len())))
```

scovich reviewed (commented):
Lots of messiness in the shredding spec here... but I think the current code is probably ok-ish, because we only support top-level variant so far? 

Main question is whether we should forbid the `AllNull` case at top level, since it's technically valid only for nested fields?

codephage2020 commented:
> This PR appears to have accumulated some conflicts; I am happy to help resolve them / get this PR merged but I wanted to check with you first

First, I sincerely apologize for the delayed response. After carefully reviewing all the feedback, I initially struggled with determining the necessary changes and how to respond—I wasn’t entirely confident in my understanding, which led to the unintended hold-up. My apologies again for the wait.

I’ve now added some comments.

Additionally, I list some comments outlining potential next steps or action items in the following, FYI.
  1. VariantArray::value() to return Option<Variant> to properly distinguish SQL NULL vs JSON null
  2. Add validation to reject top-level AllNull variants entirely
  3. Pass schema information to ShreddingState::try_new() to detect context
  4. Split AllNull handling into separate variants for top-level vs shredded contexts

Please feel free to correct or refine them—I’d greatly appreciate your guidance to ensure we move forward effectively.

alamb commented:
> Additionally, I list some comments outlining potential next steps or action items in the following, FYI.
> 
> 1. VariantArray::value() to return Option to properly distinguish SQL NULL vs JSON null

I agree with @scovich  that VariantArray::value() should continue to return `Variant` (not `Option<Variant>`) -- and the elements which where null (the validity mask was 0) `VariantArray::value()`, should return the item in the array at that location (which should be `Variant::Null` in this case, though that may not happen tody)

> 2. Add validation to reject top-level AllNull variants entirely

👍 

> 3. Pass schema information to ShreddingState::try_new() to detect context
> 4. Split AllNull handling into separate variants for top-level vs shredded contexts

I think this makes sense to me
